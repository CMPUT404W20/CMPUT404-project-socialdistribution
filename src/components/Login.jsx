import React, { Component } from "react";
import "../styles/Login.scss";
import Container from "react-bootstrap/Container";
import Row from "react-bootstrap/Row";
import Col from "react-bootstrap/Col";
import * as auth from "../services/Authentication";
import cover from "../images/cover.svg";

class Login extends Component {
  constructor(props) {
    super(props);
    this.state = {
      signup: true,
      error: "",
      username: "",
      password: "",
      passwordReentry: "",
    };
  }

  handleInputChange = (event) => {
    this.setState({
      [event.target.name]: event.target.value,
    });
  }

  handleFormSubmit = (event) => {
    event.preventDefault();

    if (!this.validateForm()) { return; }

    const { signup, username, password } = this.state;

    if (signup) {
      auth.registerUser(username, password).then((response) => {
        if (response.status === 201) {
          this.setState({
            signup: false,
            username: "",
            password: "",
          });
        }
      }).catch((err) => {
        let error = err.response.data;
        error = error.username || error.password1 || error.password2;
        this.setState({
          error: error ? error[0] : "Something went wrong",
        });
      });
    } else {
      auth.loginUser(username, password).then((response) => {
        if (response.status === 200) {
          // TODO: success: redirect to homepage
          // eslint-disable-next-line no-alert
          alert("Success");
        }
      }).catch(() => {
        this.setState({
          error: "Username or Password are incorrect",
        });
      });
    }
  }

  toggleMethod = () => {
    // toggle between sign-up and sign-in screens
    const { signup } = this.state;

    this.setState({
      signup: !signup,
      error: "",
    });
  }

  validateForm = () => {
    const {
      signup, username, password, passwordReentry,
    } = this.state;

    let error = "";
    if (!username) {
      error = "Please fill in the username";
    } else if (!password) {
      error = "Please fill in the password";
    } else if (signup) {
      if (!passwordReentry) {
        error = "Please confirm your password";
      } else if (password !== passwordReentry) {
        error = "Passwords don't match";
      } else if (password.length < 8) {
        // this is enforced by Django Auth so we need to check here
        error = "Password must be atleast 8 characters long";
      }
    }

    this.setState({
      error,
    });

    return !error;
  }

  renderForm() {
    const {
      signup, username, password, passwordReentry, error,
    } = this.state;

    let formMessage = "Let's Get Started!";
    let primaryButtonText = "Sign up";
    if (!signup) {
      formMessage = "Welcome Back!";
      primaryButtonText = "Sign in";
    }

    return (
      <form onSubmit={this.handleFormSubmit}>
        <div className="login-message">
          {formMessage}
        </div>

        <input
          type="text"
          name="username"
          className="login-field"
          placeholder="Username"
          onChange={this.handleInputChange}
          value={username}
        />

        <input
          type="password"
          name="password"
          className="login-field"
          placeholder="Password"
          onChange={this.handleInputChange}
          value={password}
        />

        {
          signup ? (
            <input
              type="password"
              name="passwordReentry"
              className="login-field"
              placeholder="Re-enter Password"
              onChange={this.handleInputChange}
              value={passwordReentry}
            />
          )
            : null
        }

        <span className="login-error-message">{error}</span>
        <button className="login-button" type="submit">{primaryButtonText}</button>
      </form>
    );
  }

  renderSecondaryOption() {
    const { signup } = this.state;

    let secondaryOptionText = "Already have account?";
    let secondaryButtonText = "Sign in";
    if (!signup) {
      secondaryOptionText = "Don't have an account?";
      secondaryButtonText = "Sign up";
    }

    return (
      <div className="login-secondary-options">
        <span className="login-secondary-options-text">{secondaryOptionText}</span>
        <button className="login-secondary-button" type="submit" onClick={this.toggleMethod}>{secondaryButtonText}</button>
      </div>
    );
  }

  render() {
    return (
      <Container fluid className="login">
        <Row>
          <Col md={7} className="login-description">
            <div className="login-heading">
              App Name
            </div>
            <div className="cover-image-wrapper">
              <img className="cover-image" src={cover} width="70%" alt="cover" />
            </div>
          </Col>
          <Col md={5} className="login-form-wrapper">
            {this.renderForm()}
            {this.renderSecondaryOption()}
          </Col>
        </Row>
      </Container>
    );
  }
}

export default Login;
