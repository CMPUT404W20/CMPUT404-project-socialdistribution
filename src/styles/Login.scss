%button {
    height: 47px;
    outline: 0;
    border-radius: 2px;
    box-shadow: 0px 0px 10px rgba(121, 121, 121, 0.25);
    font-weight: 500;
    cursor: pointer;
}

$accent-color: #4B84F2;

.login {
    background-color: grey;
    margin: 0;

    .login-description {
        background-color: #bbbbbb;
        height: 100vh;
    }

    .login-form {
        background-color: #ffffff;
        height: 100vh;
        padding: 0 50px 0;

        .login-message {
            font-size: 24px;
            font-weight: 500;
            letter-spacing: -0.05em;
            margin-bottom: 46px;
            margin-top: 30%;
        }

        .login-field {
            width: 100%;
            outline: 0;
            margin-top: 10px;
            border: 1px #EBEBEB solid;
            border-radius: 2px;
            font-size: 16px;
            padding: 15px 20px;

            &:focus {
                border: 1px #E0E0E0 solid;
            }
        }

        .login-button {
            @extend %button;
            width: 100%;
            height: 47px;
            background: $accent-color;
            color: white;
            border: 0px;
            margin-top: 76px;
        }
    }

    .login-secondary-options {
        margin-top: 50px;
        font-size: 14px;
        .login-secondary-options-text {
            line-height: 47px;
        }
    
        .login-secondary-button {
            @extend %button;
            width: 100px;
            float: right;
            border: 1px $accent-color solid;
            color: $accent-color;
        }
    }
}