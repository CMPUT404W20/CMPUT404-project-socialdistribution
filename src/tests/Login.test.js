import React from 'react';
import {shallow, mount} from 'enzyme';
import Login from '../components/Login';


describe("Login Page", () => {
  it("renders the signup flow on load", () => {
    const wrapper = shallow(<Login />);
    const loginMessage = wrapper.find(".login-message");
    expect(loginMessage.text()).toBe("Let's Get Started!");
  });

  it('should render correctly', () => {
    const component = mount(<Login />);
    expect(component).toMatchSnapshot();
  });
})

