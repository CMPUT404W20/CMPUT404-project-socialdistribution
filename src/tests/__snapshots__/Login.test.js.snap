// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login Page should render correctly 1`] = `
<Login>
  <Container
    className="login"
    fluid={true}
  >
    <div
      className="login container-fluid"
    >
      <ForwardRef
        noGutters={false}
      >
        <div
          className="row"
        >
          <Col
            className="login-description"
            md={7}
          >
            <div
              className="login-description col-md-7"
            >
              <div
                className="login-heading"
              >
                App Name
              </div>
              <div
                className="cover-image-wrapper"
              >
                <img
                  className="cover-image"
                  src="cover.svg"
                  width="70%"
                />
              </div>
            </div>
          </Col>
          <Col
            className="login-form"
            md={5}
          >
            <div
              className="login-form col-md-5"
            >
              <form
                onSubmit={[Function]}
              >
                <div
                  className="login-message"
                >
                  Let's Get Started!
                </div>
                <input
                  autoFocus={true}
                  className="login-field"
                  name="username"
                  onChange={[Function]}
                  placeholder="Username"
                  type="text"
                  value=""
                />
                <input
                  className="login-field"
                  name="password"
                  onChange={[Function]}
                  placeholder="Password"
                  type="password"
                  value=""
                />
                <input
                  className="login-field"
                  name="passwordReentry"
                  onChange={[Function]}
                  placeholder="Re-enter Password"
                  type="password"
                  value=""
                />
                <span
                  className="login-error-message"
                />
                <button
                  className="login-button"
                  type="submit"
                >
                  Sign up
                </button>
              </form>
              <div
                className="login-secondary-options"
              >
                <span
                  className="login-secondary-options-text"
                >
                  Already have account?
                </span>
                <button
                  className="login-secondary-button"
                  onClick={[Function]}
                  type="submit"
                >
                  Sign in
                </button>
              </div>
            </div>
          </Col>
        </div>
      </ForwardRef>
    </div>
  </Container>
</Login>
`;
